src/document/doc-printer.js

```diff
@@ -203,7 +203,8 @@ function fits(next, restCommands, width, options, hasLineSuffix, mustBeFlat) {
               : doc.contents,
           ]);
 
-          if (doc.id) {
+          // `doc.break = false` can't be trusted
+          if (doc.id && doc.break) {
```

Comment:
I didn't look deeply into this yet, but this doesn't look right. Groups can be broken not only because of `doc.break`. Dynamically broken groups, those that are broken because they don't fit, don't have `doc.break`.

Reply:
I know, I don't really understand this function.

> I couldn't understand why we inherit "mode" from parent group? Shouldn't we always measure width in flat mode?

This is my question.

Reply:
I tried, a) always MODE_FLAT, b) add 'undefined' mode, c) #12359. They all breaking things.

Reply:
I removed this part... Seems the same, decide `groupMode` here is wrong, even if `doc.break = true`, `fits()` printed doc is not the final output, maybe we can just remove this line?

Reply:
I guess I'm right about this, this is wrong at the first place, it happened not using ifBreak inside the group with same id. https://github.com/prettier/prettier/pull/4563/files#diff-6ffe0e63f15d2bde411262c9dd7a96802e2a5ac2dbc5f4732f03c12dce6e5b7bR313

Reply:
> I couldn't understand why we inherit "mode" from parent group? Shouldn't we always measure width in flat mode?

Is this question about this line:

```js
          const groupMode = doc.break ? MODE_BREAK : mode;
```

?

Reply:
Yes, but that.

Reply:
Consider sibling groups like this:
```js
[
  group(["(GROUP1-a", line, "GROUP1-b)"]),
  " ",
  group(["(GROUP2-a", line, "GROUP2-b)"]),
]
```
([doc explorer playground][doc1])

When the first group is printed, the second group is measured in `MODE_BREAK` inside the `fits` call. It means the group doesn't have to fit fully, only its part before the first line break. If these groups had nested groups, the same measurement rules would need to be used for them as for their parent groups. That's why `mode` is inherited.

Everything is good as long as only groups and line breaks are involved. `ifBreak` is a problem. A group measured in `MODE_BREAK` doesn't necessarily break, but `case "if-break"` in `fits` is implemented as if it were the case. This leads to [buggy results][buggy]. In fact, when `case "if-break"` is reached, it's not known yet if the group fits or breaks, so choosing between `doc.flatContents` and `doc.breakContents` can't be done fully correctly without introducing a lot of complexity (nested `fits` calls?). However, this problem somehow isn't noticeable in practice. I guess Prettier never uses `ifBreak` before line breaks inside groups, and we should consider this a known design limitation.

Similarly, your solution in this PR might be not 100% strict (`fits` doesn't really know yet whether the ancestor group with the given id breaks), but it might be good enough. I'd like to play with it a little more though.

[doc1]: https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtAOlABFg5gJwgFcAHACgxDIHEAlAeQFUAFARgFoBDdEAGiwA2ASyhx+POkzbsARgEoeAXTm9MOHlh6rseQqQo8aDFgCYuWwSLGaQk07IUhl2xXxAQSMIdADOyUJz4hADuzIEIfiicAsGcAJ5+vCAy+JxgANZwMADKnAC2cAAyVsgAZtE+YsmpGVnZJGkiuMgw+ERVcHkycAAmPb2FnFC4RJy4cABiEPh5nDBew8ggnEQwEG4AFjB5AgDqG0LwPg1gcNkRh0IAbodxS2A+iSAilfgwzKm4s2UVVQBWPgAHtkmgI4ABFIgQeA-ASVJINfCvJY9CBgdhwQEkATTOD4NwkfAiGC7IQ9GAbZAAZgAHAjCJVdqkSEtCXBXlc4G4AI5Q+AfDyRZY+diiXq9Nz4OC8oRSj5jb5IcpwqqVPJCFptVWgiF8rlK35JGCcGSk8mUpAmI2pITCYYAYQgeUVIHZAFY3ERKgAVE2RZXwkBXdoASSg-Vg2TARM8AEFw9kYHEwbDAwDAQAhGqZHL5IolA0qgC+RaAA
[buggy]: https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtAOlABFg5gJwgFcAHACgxDIHEAlAeQFUAFARgFoBDdEAGiwA2ASyhx+POkzbsARgEoeAXTm9MOHlh6rseQqQo8aDFgCYuWrEIBmAIXxxOAazIbJp2XIDUPFYJFjNEDdmM3klFUxFPhAIEhghaABnZFBOfEIAd2Y0hGSUTgEMzgBPZN4QGXxOMEc4GABlTgBbOAAZf2QrAsSxCqqauvqSapFcZBh8Il64Jpk4ABN5hdbOKFwiTlw4ADEIfCbOGHi15BBOIhgIaIALGCaBAHVroXhE4bA4etyXoQA3F+KpzAiTKIBEPXwMGYVVwB063V6ACtEgAPeqjARwACKRAg8HhAh65WG+Ahp3mEDA7DgKJIAj2cHw0RI+BEMAeQnmMGuyAAzABOYmEHoPKokU4suAQ35waIAR1x8GhsTyZ0S7FECwW0XsCqE9mhmzhSC6hN6PSaQnGk3NGOxitlJoR5RgnBkHK5PKQJhdVSEwjWAGEIE1jSApQBWaJEHoAFTdeVNRJAvymAEkoEtYPUwKy4gBBTP1GDFTEE5PIlF2aq1BrNNodJ1mgC+zaAA

Reply:
Thanks!